cpp/bench/ann/src/hnswlib/hnswlib_wrapper.h

```diff
@@ -211,12 +210,16 @@ void hnsw_lib<T>::load(const std::string& path_to_index)
 {
   if constexpr (std::is_same_v<T, float>) {
     if (metric_ == Metric::kInnerProduct) {
-      space_ = std::make_shared<hnswlib::InnerProductSpace>(dim_);
+      space_ = std::make_shared<hnswlib::InnerProductSpace<T, float>>(dim_);
     } else {
-      space_ = std::make_shared<hnswlib::L2Space>(dim_);
+      space_ = std::make_shared<hnswlib::L2Space<T, float>>(dim_);
     }
   } else if constexpr (std::is_same_v<T, uint8_t>) {
-    space_ = std::make_shared<hnswlib::L2SpaceI<T>>(dim_);
+    if (metric_ == Metric::kInnerProduct) {
+      space_ = std::make_shared<hnswlib::InnerProductSpace<T, typename hnsw_dist_t<T>::type>>(dim_);
+    } else {
+      space_ = std::make_shared<hnswlib::L2SpaceI<T>>(dim_);
+    }
```

Comment:
Can this entire if-else logic for type and metric combo now be simplified? Also, if I understand it correctly the accumulator type for `int8` and `uint8` will be `int` after your template additions in the hnswlib patch? Do we also want to do accumulations for `half` in `float`?
```suggestion
    static_assert(std::is_same_v<T, float> || std::is_same_v<T, half> || std::is_same_v<T, int8> || std::is_same_v<T, uint8>, "Only float, uint8, and int8 are supported");
    if (metric_ == Metric::kInnerProduct) {
      space_ = std::make_shared<hnswlib::InnerProductSpace<T, typename hnsw_dist_t<T>::type>>(dim_);
    } else {
      space_ = std::make_shared<hnswlib::L2SpaceI<T, typename hnsw_dist_t<T>::type>>>(dim_);
    }
```

Reply:
I will simplify it a bit by factoring the call to InnerProductSpace.
I tested the accumulations for half in float but I saw some recall drop (from over 98% to 82-92% on some cases). It would not pass the C++ tests.

Reply:
So there's no recall drop if you accumulate half in half but there is if you accumulate half in float?

Reply:
Yes. The recall drop is noticeable especially on the Inner Product distance. On the L2 distance there is no recall drop on the current tests (that go up to 2000 dim), but I added a test with 5000 dimension just to see and the recall drops to 62%, while it is still above 98% with accumulation in float.